src/factory/src/api.rs

```diff
@@ -15,8 +15,20 @@ ic_helpers::init_factory_api!(State, crate::state::get_token_bytecode());
 
 #[init]
 #[candid_method(init)]
-fn init(controller: Principal, ledger_principal: Option<Principal>) {
-    State::new(controller, ledger_principal).reset();
+pub fn init(controller: Principal, ledger_principal: Option<Principal>) {
+    State::new(controller, ledger_principal).set_global_to_self();
+}
+
+#[inspect_message]
+fn inspect_message_function() {
+    if ic_cdk::api::caller() == State::get().borrow().controller() {
```

Comment:
This will allow the controller to call `create_token` method before there's code to install. And it will actually create the canister, and then try to install empty `Vec<u8>` into it. I'm not sure if that will work in IC, though. If the `install_code` call succeeds, the canister can later be upgraded with actual code. But if it returns error, there will be an empty broken canister hanging somewhere in IC, consuming factory's cycles.

I think it's better to check here for the method being `create_token` or `upgrade`, and reject it also if the code is not there.